modules/ocf/manifests/firewall/post.pp

```diff
@@ -68,4 +70,20 @@
       before      => undef,
     }
   }
+
+
+  # Drop packets on the primary network inteface that are not whitelisted
+  # TODO: eliminate this if statement once testing is complete
+  if !$ocf::firewall::allow_other_traffic {
+    ocf::firewall::firewall46 {
+      '999 drop unrecognized input packets on primary interface':
```

Comment:
@jvperrin @abizer @chriskuehl @matthew-mcallister I've been wondering if it's a better idea to simply drop packets from DMZed servers (i.e. tsunami, werewolves, corruption) instead. It would make the mesos rules much simpler and some other rules unnecessary. What do you all think?

Reply:
In general I'd say it's better to whitelist than to blacklist. It's more work, yes, but it's possible to forget to blacklist a new public machine. Also, we would not benefit from limiting traffic from desktops, staff VMs, and DHCP addresses.

Reply:
I think that, in the rush to find all of the firewall rules we need, we may have lost sight of the original intent of this project.

My original intention with this firewall project was to place non-internal servers (i.e. servers running arbitrary user code, the DHCP range, and staff VMs) inside a DMZ. After all, we don't want malicious users on those servers attacking our internal servers, and intra-network firewalling isn't a function our existing hardware firewall can provide.

How to implement such a DMZ then? The most obvious answer is to start with one rule: internal hosts drop traffic from DMZ'ed hosts.

It doesn't take long to see that that one rule alone won't do. DMZed hosts still have to access many services on those internal hosts, such as LDAP, Kerberos, Puppet, etc. Nor would we want to prevent tsunami from accessing sites on www.ocf.berkeley.edu. Clearly, we need to whitelist those services.

Which brings us to attempt 2: on each host, we'll whitelist the ports used by actual services and drop all other traffic. Where that other traffic comes from is not considered. In this model, there is no DMZ; all hosts (except for the explicitly whitelisted supernova + hypervisors) are treated as untrusted. (This is the model we're using at this moment.)

This seemed to work for a while, but eventually problems started to show. For one, some of the rules turned out to be incredibly complicated. Take `ocf_mesos` for instance. The current list of planned rules for that module is a nightmare. There are 12 rules we need so far, and there might be even more rules needed that we haven't discovered yet. Worse, Mesos/Marathon/etc. ports are highly dynamic, which makes it difficult to find out exactly which port ranges are being used. This is a fragile, confusing, and not easily maintainable solution.

A resolution to this problem comes in the observation that not all of these ports are actually supposed to be accessible by all hosts. In particular, non-internal machines only need to access the http and https ports. If, instead of dropping traffic from all hosts by default, the internal firewalls only drop traffic from non-internal hosts by default (i.e. if non-internal hosts are placed into a DMZ), we can simplify the entire set of `ocf_mesos` firewall rules to a simple `import ocf::firewall::allow_http`. How much simpler that is! We would also similarly be able to get rid of a few other rules in other modules too.

Now, you might point out that, instead of changing our drop-traffic rule, we could instead solve our problems with `ocf_mesos` by allowing unrestricted traffic from every Mesos host + supernova + maybe desktops to every other Mesos host. That is true; however, introducing such a rule would create a discontinuity between the Mesos hosts and all other hosts. Why should Mesos hosts allow unrestricted traffic from all other Mesos hosts, when our model for every other host is to only open up a certain number of ports?

There are also conceptual considerations to take into account. Our current model is difficult to explain conceptually -- hosts firewall off exactly the set of ports that no service is listening on, so what's the point of having internal firewalls? It's also not obvious either why the internal firewalls don't render the external firewall redundant. (Eliminating the external firewall is a non-goal of this project.) On the contrary, the DMZ trust model is easy to explain: if it's not an internal server, it's in the DMZ and considered untrusted; otherwise it's trusted. The DMZ model also lends itself to a natural explanation of how this project relates to the external firewall: the external firewall is responsible for filtering traffic from outside the network, while the internal firewalls are responsible for filtering traffic only from within the network, and there is no overlap of responsibility.

It's thus evident that the DMZ model is far superior to the current model. Migrating to the DMZ model would not be difficult, as most of the rules would actually stay the same. It's in the differences, however, where the DMZ model shines.

My current proposal of implementing this DMZ involves changing the set of hosts from which traffic is dropped by default. I would use both whitelisting and blacklisting. It would be better in theory to use only whitelisting, but unfortunately our DMZ'ed hosts don't form a contiguous IP range. My current plan is roughly "blacklist `{dev-,}{tsunami,death,werewolves}`, then whitelist `169.229.226.2 - 169.229.226.139`, then blacklist every other IP in our IP block". I would appreciate discussion on how best to implement this.

Reply:
@jvperrin @abizer @chriskuehl @matthew-mcallister any more thoughts on this?

Reply:
Thanks for the long write-up on this, it definitely helps explain some of the motivations and the current state of this project.

I agree having a DMZ would work decently well and might be a good starting point, but I don't know if having just having "trusted" and "untrusted" is enough for our purposes really, because that's essentially what the current external firewall does too (our subnet is trusted, everything external is not), and this system helps with that system, but it seems to have a lot of the same shortcomings if we just have trusted and untrusted pools of hosts and some open service ports. For instance, I'd say tsunami and desktops should be untrusted in general since everyone has access to them, however something like supernova or staff VMs, although staffers can have arbitrary services, should be somewhat more trusted than tsunami (allowing SSH to other servers for instance, but maybe nothing else for staff VMs?), so I feel like multiple levels of trust would make this system more flexible now and in the future. Also, I think having all internal servers open to each other makes that perhaps too open, although I do like the plan of making the rules simpler than they currently are if possible. I agree that splitting things up more granularly like this makes things more complicated, but I'm worried otherwise we won't be able to express the full complexity of our setup of different hosts if we just separate hosts into two groups.